jetbrains: idea-oss, pycharm-oss: 2025.1.1.1 -> 2025.3.1

theCapypara opened the pull request:
Updates `idea-oss` and `pycharm-oss` to the latest version. Fixes various build issues & also automates generating the jps & restarter hashes.

The source build is now also always updated from the latest release info, instead of just using whatever the bin versions use. This means the build processes are now independent.

There's quite a bit of code duplication between the bin/source update scripts. I plan to refactor this eventually. After this PR I'll introduce support for `updateScript` if & when #470147 is also merged.

IMO we should exceptionally backport this to 25.11 since we haven't touched the source versions in a long time & to get the fixes onto the stable branch. Also since we are going to backport #466331 as well.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [ ] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

theCapypara committed:
jetbrains: fix source update script

theCapypara committed:
jetbrains: update basic source dependencies and patches

Co-Authored-By: Matt Sturgeon <matt@sturgeon.me.uk>

theCapypara committed:
jetbrains: source builds: auto-update jps & restarter hashes

theCapypara committed:
jetbrains: source builds: fetch latest versions

theCapypara committed:
jetbrains: source builds: fixed & hardened maven deps update script

MattSturgeon commented on pkgs/applications/editors/jetbrains/source/update.py:
```diff
@@ -80,6 +81,33 @@ def prefetch_android(variant: str, buildNumber: str) -> str:
+            jpsHash = "sha256-AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=";
+        }}
+    '''], capture_output=True, check=True, text=True)
+    return convert_hash_to_sri(output.stdout.strip())
```
Same.

MattSturgeon reviewed (commented):
Lot's of code here I'm not too familiar with, but looking at the diff things seem ok.

MattSturgeon commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review-gha`](https://github.com/Defelo/nixpkgs-review-gha)

Command: `nixpkgs-review pr 470774`
Commit: [`c3aa3d92ab87754d733141b3520d582c88f2c5f5`](https://github.com/NixOS/nixpkgs/commit/c3aa3d92ab87754d733141b3520d582c88f2c5f5) ([subsequent changes](https://github.com/NixOS/nixpkgs/compare/c3aa3d92ab87754d733141b3520d582c88f2c5f5..pull/470774/head))
Merge: [`85f1289cc9301f88b18d5f5e929d0af77e620bd5`](https://github.com/NixOS/nixpkgs/commit/85f1289cc9301f88b18d5f5e929d0af77e620bd5)

Logs: https://github.com/MattSturgeon/nixpkgs-review-gha/actions/runs/20317520617


---
### `x86_64-linux`
<details>
  <summary>:x: 2 packages failed to build:</summary>
  <ul>
    <li>jetbrains.idea-oss</li>
    <li>jetbrains.pycharm-oss</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 13 packages built:</summary>
  <ul>
    <li>jetbrains.clion</li>
    <li>jetbrains.datagrip</li>
    <li>jetbrains.dataspell</li>
    <li>jetbrains.gateway</li>
    <li>jetbrains.goland</li>
    <li>jetbrains.idea</li>
    <li>jetbrains.mps</li>
    <li>jetbrains.phpstorm</li>
    <li>jetbrains.pycharm</li>
    <li>jetbrains.rider</li>
    <li>jetbrains.ruby-mine</li>
    <li>jetbrains.rust-rover</li>
    <li>jetbrains.webstorm</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:x: 2 packages failed to build:</summary>
  <ul>
    <li>jetbrains.idea-oss</li>
    <li>jetbrains.pycharm-oss</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 13 packages built:</summary>
  <ul>
    <li>jetbrains.clion</li>
    <li>jetbrains.datagrip</li>
    <li>jetbrains.dataspell</li>
    <li>jetbrains.gateway</li>
    <li>jetbrains.goland</li>
    <li>jetbrains.idea</li>
    <li>jetbrains.mps</li>
    <li>jetbrains.phpstorm</li>
    <li>jetbrains.pycharm</li>
    <li>jetbrains.rider</li>
    <li>jetbrains.ruby-mine</li>
    <li>jetbrains.rust-rover</li>
    <li>jetbrains.webstorm</li>
  </ul>
</details>

theCapypara commented:
@MattSturgeon Can you try again? Might have been a temporary issue on Jetbrain's end. The URLs in your build logs (e. g. https://cache-redirector.jetbrains.com/packages.jetbrains.team/maven/p/grazi/grazie-platform-public/ai/grazie/utils/utils-ktor-jvm/0.8.74/utils-ktor-jvm-0.8.74.jar) are working and I can build everything just fine.

I think that service is just a mirror for maven repos and some other pages, we could probably just add the non-mirrored URLs as well, but that'd require some more changes here. Currently fallback URLs are not used. I'll write that down as another future improvement to do.

For example for the above URL https://packages.jetbrains.team/maven/p/grazi/grazie-platform-public/ai/grazie/utils/utils-ktor-jvm/0.8.74/utils-ktor-jvm-0.8.74.jar works just fine as well.

There were [some URLs](https://github.com/NixOS/nixpkgs/pull/470774/changes#diff-7cc785305b2622c6737fee972f6502ef9ae8ef80fb756c48afd72c6710abe17aL201-R184) I noticed changed location and I replaced them with the new one used. But those would actually probably also work via the cache redirector... I don't know much about it's purpose - or anything really.

theCapypara commented:
That's really odd. I have no idea why it doesn't build inside Github Actions.

theCapypara committed:
jetbrains.*-oss: Add fallback URLs w/o the cache redirector

Co-Authored-By: Matt Sturgeon <matt@sturgeon.me.uk>

theCapypara committed:
jetbrains.*-oss: Do not use hardcoded build dir paths

Co-Authored-By: Matt Sturgeon <matt@sturgeon.me.uk>

theCapypara committed:
jetbrains.*-oss: replace brokenPlugins.json with an empty list

MattSturgeon commented on pkgs/applications/editors/jetbrains/source/build.nix:
```diff
@@ -190,33 +174,35 @@ let
+          "https://cache-redirector.jetbrains.com/${url}/${entry.url}"
+          "https://${url}/${entry.url}"
+        ]) repositories
+      );
```
In this case, I think you can `concatMap` instead of `flatten (map)`.

In other cases, it may be better to reach for `concatLists` before resorting to `flatten`

```suggestion
      urls = lib.concatMap (url: [
        "https://cache-redirector.jetbrains.com/${url}/${entry.url}"
        "https://${url}/${entry.url}"
      ]) repositories;
```

theCapypara commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review-gha`](https://github.com/Defelo/nixpkgs-review-gha)

Command: `nixpkgs-review pr 470774`
Commit: [`ddd6b993a4c01c564f2fcde57dc74fc1980c332c`](https://github.com/NixOS/nixpkgs/commit/ddd6b993a4c01c564f2fcde57dc74fc1980c332c) ([subsequent changes](https://github.com/NixOS/nixpkgs/compare/ddd6b993a4c01c564f2fcde57dc74fc1980c332c..pull/470774/head))
Merge: [`6f229c6f92dbe054dca4c237157c24216ec51c21`](https://github.com/NixOS/nixpkgs/commit/6f229c6f92dbe054dca4c237157c24216ec51c21)

Logs: https://github.com/theCapypara/nixpkgs-review-gha/actions/runs/20446566183


---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 15 packages built:</summary>
  <ul>
    <li>jetbrains.clion</li>
    <li>jetbrains.datagrip</li>
    <li>jetbrains.dataspell</li>
    <li>jetbrains.gateway</li>
    <li>jetbrains.goland</li>
    <li>jetbrains.idea</li>
    <li>jetbrains.idea-oss</li>
    <li>jetbrains.mps</li>
    <li>jetbrains.phpstorm</li>
    <li>jetbrains.pycharm</li>
    <li>jetbrains.pycharm-oss</li>
    <li>jetbrains.rider</li>
    <li>jetbrains.ruby-mine</li>
    <li>jetbrains.rust-rover</li>
    <li>jetbrains.webstorm</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:white_check_mark: 15 packages built:</summary>
  <ul>
    <li>jetbrains.clion</li>
    <li>jetbrains.datagrip</li>
    <li>jetbrains.dataspell</li>
    <li>jetbrains.gateway</li>
    <li>jetbrains.goland</li>
    <li>jetbrains.idea</li>
    <li>jetbrains.idea-oss</li>
    <li>jetbrains.mps</li>
    <li>jetbrains.phpstorm</li>
    <li>jetbrains.pycharm</li>
    <li>jetbrains.pycharm-oss</li>
    <li>jetbrains.rider</li>
    <li>jetbrains.ruby-mine</li>
    <li>jetbrains.rust-rover</li>
    <li>jetbrains.webstorm</li>
  </ul>
</details>

---
### `x86_64-darwin`
:white_check_mark: *No rebuilds*

---
### `aarch64-darwin`
:white_check_mark: *No rebuilds*

MattSturgeon commented on pkgs/applications/editors/jetbrains/source/build.nix:
```diff
@@ -261,6 +245,9 @@ stdenvNoCC.mkDerivation rec {
 
+  # Regarding brokenPlugins.json:
+  #  We provide an empty brokenPlugins.json because the original file is 2.8 MB in size and can't be reliably
+  #  downloaded from archive.org without hash changes. The file is not important for the IDEs to function properly.
```
Not sure this needs to explicitly mention archive.org?


```suggestion
  #  downloaded without hash changes. The file is not important for the IDEs to function properly.
```

MattSturgeon commented on pkgs/applications/editors/jetbrains/source/build.nix:
```diff
@@ -300,7 +287,7 @@ stdenvNoCC.mkDerivation rec {
       -Dintellij.build.unix.snaps=false \
       --java-argfile-target=java_argfile \
-      /build/source \
+      $(pwd) \
```
same

MattSturgeon reviewed (commented):
SGTM. Great to see nixpkgs-review passing, well done!

Nits, ideas, questions, etc:

(I'll caveat that I'm not really familiar with the "big picture" of most of this code, so my review is focused on intricacies rather than the overall design)

ShadowRZ reviewed (approved):
LGTM

@aoli-al I personally think `jetbrains.jdk-17` should be dropped too.

theCapypara commented on pkgs/applications/editors/jetbrains/source/build.nix:
```diff
@@ -169,16 +152,17 @@ let
       runHook preBuild
 
-      ant -Duser.home=${jpsRepo} -Dbuild.dir=/build/out -f jps-bootstrap-classpath.xml
+      BUILD_HOME=$(mktemp -d)
```
I think there wasn't ever really a reason to build this inside the build dir. A tempdir for this is fine, it doesn't really matter where we build. So if this increases compatibility with some systems I'd say it's fine.

theCapypara commented on pkgs/applications/editors/jetbrains/readme.md:
```diff
@@ -24,14 +24,12 @@ The jdk is in `pkgs/development/compilers/jetbrains-jdk`.
+   - Make sure the Kotlin version used is correct.
    - Do a test build
-     - Notice that sometimes a newer Kotlin version is required to build from source, if build fails, first check the recommended Kotlin version in `.idea/kotlinc.xml` in the IDEA source root
-     - Feel free to update the Kotlin version to a compatible one
```
Hm, I may have shortened this a bit too much, I'll add the info about the `.idea/kotlinc.xml` back.

theCapypara commented:
@MattSturgeon I applied your suggestions in a new commit (since I have a hard time trying to squeeze it into the existing commits).

I hope that's ok, otherwise I can try going through it piece-by-piece.

theCapypara commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review-gha`](https://github.com/Defelo/nixpkgs-review-gha)

Command: `nixpkgs-review pr 470774`
Commit: [`575a1455a3973a1d45426bcff91d531b14b49fc2`](https://github.com/NixOS/nixpkgs/commit/575a1455a3973a1d45426bcff91d531b14b49fc2) ([subsequent changes](https://github.com/NixOS/nixpkgs/compare/575a1455a3973a1d45426bcff91d531b14b49fc2..pull/470774/head))
Merge: [`b2918ddd80da1ba9ccabaadd945b62e1b6de6bd5`](https://github.com/NixOS/nixpkgs/commit/b2918ddd80da1ba9ccabaadd945b62e1b6de6bd5)

Logs: https://github.com/theCapypara/nixpkgs-review-gha/actions/runs/20483931435


---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 15 packages built:</summary>
  <ul>
    <li>jetbrains.clion</li>
    <li>jetbrains.datagrip</li>
    <li>jetbrains.dataspell</li>
    <li>jetbrains.gateway</li>
    <li>jetbrains.goland</li>
    <li>jetbrains.idea</li>
    <li>jetbrains.idea-oss</li>
    <li>jetbrains.mps</li>
    <li>jetbrains.phpstorm</li>
    <li>jetbrains.pycharm</li>
    <li>jetbrains.pycharm-oss</li>
    <li>jetbrains.rider</li>
    <li>jetbrains.ruby-mine</li>
    <li>jetbrains.rust-rover</li>
    <li>jetbrains.webstorm</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:white_check_mark: 15 packages built:</summary>
  <ul>
    <li>jetbrains.clion</li>
    <li>jetbrains.datagrip</li>
    <li>jetbrains.dataspell</li>
    <li>jetbrains.gateway</li>
    <li>jetbrains.goland</li>
    <li>jetbrains.idea</li>
    <li>jetbrains.idea-oss</li>
    <li>jetbrains.mps</li>
    <li>jetbrains.phpstorm</li>
    <li>jetbrains.pycharm</li>
    <li>jetbrains.pycharm-oss</li>
    <li>jetbrains.rider</li>
    <li>jetbrains.ruby-mine</li>
    <li>jetbrains.rust-rover</li>
    <li>jetbrains.webstorm</li>
  </ul>
</details>

---
### `x86_64-darwin`
:white_check_mark: *No rebuilds*

---
### `aarch64-darwin`
:white_check_mark: *No rebuilds*

MattSturgeon reviewed (commented):
A couple non-blocking bash hygine nits, in case you weren't aware of the subtleties here. Up to you whether to action them.

Overall PR still SGTM.

theCapypara committed:
jetbrains: source build updates: write repositories without prefix

theCapypara committed:
jetbrains: idea-oss, pycharm-oss: 2025.1.1.1 -> 2025.3.1

theCapypara commented:
I was able to squash the review feedback into previous commits.

And:
- I added one more commit that makes sure that the `source/update.py` writes the `source/sources.json` `repositories` entries without `https://` or `https://cache-redirector.jetbrains.com/` prefixes like I manually edited them; almost missed that.
- I updated the IDEs to `2025.3.1` and remove the previous update commit (`d8dc15032231 jetbrains: idea-oss, pycharm-oss: 2025.1.1.1 -> 2025.3`) from this PR.

Eveeifyeve commented:
> > Please credit the suggestions provided by the reviewers via `Co-authored-by` so we are not breaking licensing.
> 
> While `Co-authored-by` metadata can be nice to have, I don't think it's a licensing requirement.
> 
> Git commit metadata isn't the authoritative source of copyright in nixpkgs; licensing is covered by [`COPYING`](https://github.com/NixOS/nixpkgs/blob/master/COPYING?rgh-link-date=2025-12-25T11%3A29%3A21Z), which attributes copyright collectively to:
> 
> > Copyright (c) 2003-2025 Eelco Dolstra and the Nixpkgs/NixOS contributors
> 
> Regardless, I'm fine for my suggestions to be merged into Nixpkgs without attribution.
> 
> TL;DR: Adding `Co-authored-by` can be good practice, but it isn't required here for licensing reasons.

I thought it was, because a long time ago @emilazy said I needed to do it so I am confused now so it's not required?????
